src/main/scala/quantiles/QuantileNonSample.scala

```diff
@@ -0,0 +1,252 @@
+package quantiles
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+import scala.util.control.Breaks._
+import scala.util.Sorting.quickSort
+
+class QuantileNonSample[T](val sketchSize: Int,
+                           private val shrinkingFactor: Double = 0.64)
+                          (implicit ordering: Ordering[T],
+                           ct: ClassTag[T]) extends Serializable{
+
+  private var curNumOfCompactors = 0
+  // initialize with ArrayBuffer, add compactors later
+  private var compactors = ArrayBuffer[NonSampleCompactor[T]]()
+  // number of items in compactors
+  private var compactorActualSize = 0
+  // overall capacity of compactors
+  private var compactorTotalSize = 0
+  expand()
+
+
+
+  // expand a layer of compactor
+  private def expand(): Unit = {
+    compactors = compactors :+ new NonSampleCompactor[T]
+    curNumOfCompactors = compactors.length
+    var size = 0
+    (0 until curNumOfCompactors).foreach { height =>
+      size = size + capacity(height)
+    }
+    compactorTotalSize = size
+  }
+
+  private def capacity(height:Int): Int = {
+    Math.ceil(sketchSize * Math.pow(shrinkingFactor, height)).toInt + 1
+  }
+
+  /**
+    * update the sketch with a single item
+    *
+    * @param item new item observed by the sketch
+    */
+  def update(item: T): Unit = {
+    compactors(0).buffer = compactors(0).buffer :+ item
+
+    compactorActualSize = compactorActualSize + 1
+    if (compactorActualSize > compactorTotalSize) {
+      condense()
+    }
+  }
+
+  private def condense(): Unit = {
+    breakable {
+      for (height <- compactors.indices) {
+        if (compactors(height).buffer.length >= capacity(height)) {
+          if (height + 1 >= curNumOfCompactors) expand()
+          val output: Array[T] = compactors(height).compact
+          output.foreach(element => compactors(height + 1).buffer = compactors(height + 1).buffer :+ element)
+          var size = 0
+          (0 until curNumOfCompactors).foreach { height =>
+            size = size + this.compactors(height).buffer.length
+          }
+          compactorActualSize = size
+          // implement lazy here
+          break
+        }
+      }
+    }
+  }
+
+  private def compare[T](o1: T, o2: T)(implicit ord: Ordering[T]): Boolean = ord.gt(o1, o2)
+
+  /**
+    * Get the map which contains the rank of all items which is currently in the sketch.
+    * @return the map which contains the rank of all items which is currently in the sketch (RankMap)
+    */
+  def getRankMap(): mutable.Map[T,Long] = {
+    val Set = output.toMap.keySet
+    var states = scala.collection.mutable.Map[T,Long]()
+    Set.foreach{item =>
+      var curWeight = 0L
+      output.foreach(tuple => {
+        if (!compare(tuple._1,item)) {
```

Comment:
compare returns a sign based on the comparison (see 
https://www.scala-lang.org/api/current/scala/math/Ordering$$Int$.html#compare(x:Int,y:Int):Int  )
comapre(a,b) = 
   -1 if a<b 
   0 if a==b
    1 if a > b
This means that !compare() will be true only if the numbers are equal. This is not the right behaviour

Should be 
if (compare(tuple._1, item) <= 0 )

Reply:
The compare function I have used here is a private function that I have defined just above the getRankMap(), not the compare function listed in the link.
private def compare[T](o1: T, o2: T)(implicit ord: Ordering[T]): Boolean = ord.gt(o1, o2)
which will only return true if the first element is larger than the second.
But should I refactor the function name to avoid confusions like this?

Reply:
Sorry - I missed this. Probably best to refactor and call it greaterThan

Reply:
Sure. I have refactored it, thank you for the suggestion!